File: src/parser/unit.zig
const std = @import("std");
const ast = @import("../ast/mod.zig");
const Parser = @import("parser.zig").Parser;
const parseMap = @import("map.zig").parseMap;

fn parseStmt(self: *Parser, body: *std.ArrayList(ast.Stmt)) Parser.ParseError!void {
    if (self.match(.keyword_reg)) {
        const var_loc = self.current.loc;
        const var_name_tok = try self.expect(.identifier);
        _ = try self.expect(.equals);
        const value_expr = try self.parseExpr();
        _ = try self.expect(.semicolon);

        try body.append(self.allocator, .{
            .kind = .{ .VarDecl = .{
                .name = try self.allocator.dupe(u8, var_name_tok.lexeme),
                .var_type = .reg,
                .value = value_expr,
            } },
            .loc = var_loc,
        });
        return;
    }

    if (self.match(.keyword_imm)) {
        const var_loc = self.current.loc;
        const var_name_tok = try self.expect(.identifier);
        _ = try self.expect(.equals);
        const value_expr = try self.parseExpr();
        _ = try self.expect(.semicolon);

        try body.append(self.allocator, .{
            .kind = .{ .VarDecl = .{
                .name = try self.allocator.dupe(u8, var_name_tok.lexeme),
                .var_type = .imm,
                .value = value_expr,
            } },
            .loc = var_loc,
        });
        return;
    }

    if (self.match(.keyword_heap)) {
        const var_loc = self.current.loc;
        const var_name_tok = try self.expect(.identifier);
        _ = try self.expect(.equals);

        const map_name_tok = try self.expect(.identifier);
        _ = try self.expect(.dot);
        const lookup_tok = try self.expect(.identifier);
        if (!std.mem.eql(u8, lookup_tok.lexeme, "lookup")) {
            return self.parseError("expected 'lookup'");
        }

        _ = try self.expect(.l_paren);
        const key_expr = try self.parseExpr();
        _ = try self.expect(.r_paren);
        _ = try self.expect(.semicolon);

        try body.append(self.allocator, .{
            .kind = .{ .HeapVarDecl = .{
                .name = try self.allocator.dupe(u8, var_name_tok.lexeme),
                .lookup = .{
                    .map_name = try self.allocator.dupe(u8, map_name_tok.lexeme),
                    .key_expr = key_expr,
                },
            } },
            .loc = var_loc,
        });
        return;
    }

    if (self.match(.keyword_return)) {
        const return_loc = self.current.loc;
        const v = try self.parseExpr();
        _ = try self.expect(.semicolon);

        try body.append(self.allocator, .{
            .kind = .{ .Return = v },
            .loc = return_loc,
        });
        return;
    }

    if (self.match(.keyword_if)) {
        const if_loc = self.current.loc;
        _ = try self.expect(.keyword_guard);
        _ = try self.expect(.l_paren);
        const var_tok = try self.expect(.identifier);
        _ = try self.expect(.r_paren);
        _ = try self.expect(.l_brace);

        var guard_body = try std.ArrayList(ast.Stmt).initCapacity(self.allocator, 0);
        while (!self.match(.r_brace)) {
            try parseStmt(self, &guard_body);
        }

        try body.append(self.allocator, .{
            .kind = .{ .IfGuard = .{
                .condition = .{
                    .kind = .{ .VarRef = try self.allocator.dupe(u8, var_tok.lexeme) },
                    .loc = var_tok.loc,
                },
                .body = try guard_body.toOwnedSlice(self.allocator),
            } },
            .loc = if_loc,
        });
        return;
    }

    const target = try self.parseExpr();
    if (self.match(.equals)) {
        const value = try self.parseExpr();
        _ = try self.expect(.semicolon);
        try body.append(self.allocator, .{
            .kind = .{ .Assignment = .{
                .target = target,
                .op = "=",
                .value = value,
            } },
            .loc = target.loc,
        });
        return;
    } else if (self.match(.plus_equals)) {
        const value = try self.parseExpr();
        _ = try self.expect(.semicolon);
        try body.append(self.allocator, .{
            .kind = .{ .Assignment = .{
                .target = target,
                .op = "+=",
                .value = value,
            } },
            .loc = target.loc,
        });
        std.debug.print("[DEBUG] Parsed assignment\n", .{});
        return;
    }

    return self.parseError("unexpected statement");
}

pub fn parseExpr(self: *Parser) Parser.ParseError!*ast.Expr {
    if (self.check(.star)) {
        _ = try self.expect(.star);
        const inner = try self.parseExpr();
        const expr = try self.allocator.create(ast.Expr);
        expr.* = .{
            .kind = .{ .Dereference = inner },
            .loc = inner.loc,
        };
        return expr;
    }

    if (self.check(.number)) {
        const num_tok = try self.expect(.number);
        const expr = try self.allocator.create(ast.Expr);
        expr.* = .{
            .kind = .{ .Number = num_tok.int_value },
            .loc = num_tok.loc,
        };
        return expr;
    }

    const receiver_tok = try self.expect(.identifier);

    if (self.match(.dot)) {
        const method_tok = try self.expect(.identifier);
        _ = try self.expect(.l_paren);
        const arg = try self.parseExpr();
        _ = try self.expect(.r_paren);

        const expr = try self.allocator.create(ast.Expr);
        expr.* = .{
            .kind = .{ .MethodCall = .{
                .receiver = try self.allocator.dupe(u8, receiver_tok.lexeme),
                .method = try self.allocator.dupe(u8, method_tok.lexeme),
                .arg = arg,
            } },
            .loc = receiver_tok.loc,
        };
        return expr;
    } else {
        const expr = try self.allocator.create(ast.Expr);
        expr.* = .{
            .kind = .{ .VarRef = try self.allocator.dupe(u8, receiver_tok.lexeme) },
            .loc = receiver_tok.loc,
        };
        return expr;
    }
}

pub fn parseUnit(self: *Parser) Parser.ParseError!ast.Unit {
    const unit_loc = self.current.loc;
    _ = try self.expect(.keyword_unit);

    const name_tok = try self.expect(.identifier);
    std.debug.print("[DEBUG] Parsing unit: {s}\n", .{name_tok.lexeme});
    _ = try self.expect(.l_brace);

    var sections = try std.ArrayList([]const u8).initCapacity(self.allocator, 0);
    var body = try std.ArrayList(ast.Stmt).initCapacity(self.allocator, 0);
    var license: ?[]const u8 = null;

    while (!self.match(.r_brace)) {
        if (self.match(.keyword_section)) {
            _ = try self.expect(.colon);
            const s = try self.expect(.string_literal);
            const txt = if (s.lexeme.len >= 2) s.lexeme[1 .. s.lexeme.len - 1] else "";
            try sections.append(self.allocator, try self.allocator.dupe(u8, txt));
            _ = try self.expect(.semicolon);
            std.debug.print("[DEBUG] Parsed section\n", .{});
            continue;
        }

        if (self.match(.keyword_license)) {
            _ = try self.expect(.colon);
            const s = try self.expect(.string_literal);
            const txt = if (s.lexeme.len >= 2) s.lexeme[1 .. s.lexeme.len - 1] else "";
            license = try self.allocator.dupe(u8, txt);
            _ = try self.expect(.semicolon);
            std.debug.print("[DEBUG] Parsed license\n", .{});
            continue;
        }

        try parseStmt(self, &body);
    }

    std.debug.print("[DEBUG] Finished parsing unit\n", .{});

    return ast.Unit{
        .name = try self.allocator.dupe(u8, name_tok.lexeme),
        .loc = unit_loc,
        .sections = try sections.toOwnedSlice(self.allocator),
        .license = license,
        .body = try body.toOwnedSlice(self.allocator),
    };
}

File: src/parser/unit.rs
use super::{Parser, ParseError};
use crate::{ast::{
    Assignment, AssignmentOp,  Expr, ExprKind, HeapLookup, HeapVarDecl, IfGuard, MethodCall, Stmt, StmtKind, Unit, VarDecl, VarType
}, parser::{TokenKind, map::expect_token}};
use std::boxed::Box;

pub fn parse_unit(parser: &mut Parser) -> Result<Unit, ParseError> {
    let unit_loc = parser.current_loc();
    expect_token(parser, TokenKind::KeywordUnit)?;

    let name_tok = parser.expect(TokenKind::Identifier)?;
    expect_token(parser, TokenKind::LBrace)?;

    let mut sections = Vec::new();
    let mut body = Vec::new();
    let mut license: Option<String> = None;

    while !parser.check(TokenKind::RBrace) {
        if parser.r#match(TokenKind::KeywordSection) {
            expect_token(parser, TokenKind::Colon)?;
            let s = parser.expect(TokenKind::StringLiteral)?;
            
            let txt = s.lexeme.trim_matches('"').to_string();
            sections.push(txt);
            expect_token(parser, TokenKind::Semicolon)?;
            continue;
        }

        if parser.r#match(TokenKind::KeywordLicense) {
            expect_token(parser, TokenKind::Colon)?;
            let s = parser.expect(TokenKind::StringLiteral)?;
            
            let txt = s.lexeme.trim_matches('"').to_string();
            license = Some(txt);
            expect_token(parser, TokenKind::Semicolon)?;
            continue;
        }

        parse_stmt(parser, &mut body)?;
    }

    expect_token(parser, TokenKind::RBrace)?;

    Ok(Unit {
        name: name_tok.lexeme,
        loc: unit_loc,
        sections,
        license,
        body,
    })
}

fn parse_stmt(parser: &mut Parser, body: &mut Vec<Stmt>) -> Result<(), ParseError> {
    if parser.r#match(TokenKind::KeywordReg) {
        let var_loc = parser.current_loc();
        let var_name_tok = parser.expect(TokenKind::Identifier)?;
        expect_token(parser, TokenKind::Equals)?;
        let value_expr = parse_expr(parser)?;
        expect_token(parser, TokenKind::Semicolon)?;

        body.push(Stmt {
            kind: StmtKind::VarDecl(VarDecl {
                name: var_name_tok.lexeme,
                var_type: VarType::Reg,
                value: Box::new(value_expr),
            }),
            loc: var_loc,
        });
        return Ok(());
    }

    if parser.r#match(TokenKind::KeywordImm) {
        let var_loc = parser.current_loc();
        let var_name_tok = parser.expect(TokenKind::Identifier)?;
        expect_token(parser, TokenKind::Equals)?;
        let value_expr = parse_expr(parser)?;
        expect_token(parser, TokenKind::Semicolon)?;

        body.push(Stmt {
            kind: StmtKind::VarDecl(VarDecl {
                name: var_name_tok.lexeme,
                var_type: VarType::Imm,
                value: Box::new(value_expr),
            }),
            loc: var_loc,
        });
        return Ok(());
    }

    if parser.r#match(TokenKind::KeywordHeap) {
        let var_loc = parser.current_loc();
        let var_name_tok = parser.expect(TokenKind::Identifier)?;
        expect_token(parser, TokenKind::Equals)?;

        let map_name_tok = parser.expect(TokenKind::Identifier)?;
        expect_token(parser, TokenKind::Dot)?;
        let lookup_tok = parser.expect(TokenKind::Identifier)?;
        
        if lookup_tok.lexeme != "lookup" {
            return Err(parser.error_with_help(
                "Expected 'lookup' method",
                format!("Found: {}", lookup_tok.lexeme)
            ));
        }

        expect_token(parser, TokenKind::LParen)?;
        let key_expr = parse_expr(parser)?;
        expect_token(parser, TokenKind::RParen)?;
        expect_token(parser, TokenKind::Semicolon)?;

        body.push(Stmt {
            kind: StmtKind::HeapVarDecl(HeapVarDecl {
                name: var_name_tok.lexeme,
                lookup: HeapLookup {
                    map_name: map_name_tok.lexeme,
                    key_expr: Box::new(key_expr),
                },
            }),
            loc: var_loc,
        });
        return Ok(());
    }

    if parser.r#match(TokenKind::KeywordReturn) {
        let return_loc = parser.current_loc();
        let v = parse_expr(parser)?;
        expect_token(parser, TokenKind::Semicolon)?;

        body.push(Stmt {
            kind: StmtKind::Return(Box::new(v)),
            loc: return_loc,
        });
        return Ok(());
    }

    if parser.r#match(TokenKind::KeywordIf) {
        let if_loc = parser.current_loc();
        expect_token(parser, TokenKind::KeywordGuard)?;
        expect_token(parser, TokenKind::LParen)?;
        let var_tok = parser.expect(TokenKind::Identifier)?;
        expect_token(parser, TokenKind::RParen)?;
        expect_token(parser, TokenKind::LBrace)?;

        let mut guard_body = Vec::new();
        while !parser.r#match(TokenKind::RBrace) {
            parse_stmt(parser, &mut guard_body)?;
        }

        body.push(Stmt {
            kind: StmtKind::IfGuard(IfGuard {
                condition: Expr {
                    kind: ExprKind::Variable(var_tok.lexeme.clone()),
                    loc: var_tok.loc,
                },
                body: guard_body,
            }),
            loc: if_loc,
        });
        return Ok(());
    }

    let target = parse_expr(parser)?;
    let target_loc = target.loc;
    if parser.r#match(TokenKind::Equals) {
        let value = parse_expr(parser)?;
        expect_token(parser, TokenKind::Semicolon)?;
        
        body.push(Stmt {
            kind: StmtKind::Assignment(Assignment {
                target: Box::new(target),
                op: AssignmentOp::Assign,
                value: Box::new(value),
            }),
            loc: target_loc,
        });
        return Ok(());
    } else if parser.r#match(TokenKind::PlusEquals) {
        let value = parse_expr(parser)?;
        expect_token(parser, TokenKind::Semicolon)?;
        
        body.push(Stmt {
            kind: StmtKind::Assignment(Assignment {
                target: Box::new(target),
                op: AssignmentOp::AddAssign,
                value: Box::new(value),
            }),
            loc: target_loc,
        });
        return Ok(());
    }

    Err(parser.error("Unexpected statement")
        .with_help("Expected: reg, imm, heap, return, if, or expression"))
}

pub fn parse_expr(parser: &mut Parser) -> Result<Expr, ParseError> {
    parse_add(parser)
}

fn parse_add(parser: &mut Parser) -> Result<Expr, ParseError> {
    let mut expr = parse_mul(parser)?;

    loop {
        if parser.r#match(TokenKind::Plus) {
            let rhs = parse_mul(parser)?;
            let loc = expr.loc;
            expr = Expr {
                kind: ExprKind::Binary(crate::ast::BinaryExpr {
                    op: crate::ast::BinOp::Add,
                    left: Box::new(expr),
                    right: Box::new(rhs),
                }),
                loc,
            };
            continue;
        }

        if parser.r#match(TokenKind::Minus) {
            let rhs = parse_mul(parser)?;
            let loc = expr.loc;
            expr = Expr {
                kind: ExprKind::Binary(crate::ast::BinaryExpr {
                    op: crate::ast::BinOp::Sub,
                    left: Box::new(expr),
                    right: Box::new(rhs),
                }),
                loc,
            };
            continue;
        }

        break;
    }

    Ok(expr)
}

// * / %
fn parse_mul(parser: &mut Parser) -> Result<Expr, ParseError> {
    let mut expr = parse_unary(parser)?;

    loop {
        if parser.r#match(TokenKind::Star) {
            let rhs = parse_unary(parser)?;
            let loc = expr.loc;
            expr = Expr {
                kind: ExprKind::Binary(crate::ast::BinaryExpr {
                    op: crate::ast::BinOp::Mul,
                    left: Box::new(expr),
                    right: Box::new(rhs),
                }),
                loc,
            };
            continue;
        }

        if parser.r#match(TokenKind::Slash) {
            let rhs = parse_unary(parser)?;
            let loc = expr.loc;
            expr = Expr {
                kind: ExprKind::Binary(crate::ast::BinaryExpr {
                    op: crate::ast::BinOp::Div,
                    left: Box::new(expr),
                    right: Box::new(rhs),
                }),
                loc,
            };
            continue;
        }

        if parser.r#match(TokenKind::Percent) {
            let rhs = parse_unary(parser)?;
            let loc = expr.loc;
            expr = Expr {
                kind: ExprKind::Binary(crate::ast::BinaryExpr {
                    op: crate::ast::BinOp::Mod,
                    left: Box::new(expr),
                    right: Box::new(rhs),
                }),
                loc,
            };
            continue;
        }

        break;
    }

    Ok(expr)
}

// unary: *expr (dereference)
fn parse_unary(parser: &mut Parser) -> Result<Expr, ParseError> {
    if parser.r#match(TokenKind::Star) {
        let inner = parse_unary(parser)?;
        let inner_loc = inner.loc;
        return Ok(Expr {
            kind: ExprKind::Dereference(Box::new(inner)),
            loc: inner_loc,
        });
    }

    parse_primary(parser)
}

fn parse_primary(parser: &mut Parser) -> Result<Expr, ParseError> {
    // number
    if parser.check(TokenKind::Number) {
        let num_tok = parser.expect(TokenKind::Number)?;
        let value = num_tok
            .int_value
            .ok_or_else(|| parser.error("Invalid number literal"))?;

        return Ok(Expr {
            kind: ExprKind::Number(value),
            loc: num_tok.loc,
        });
    }

    // (expr)
    if parser.r#match(TokenKind::LParen) {
        let e = parse_expr(parser)?;
        expect_token(parser, TokenKind::RParen)?;
        return Ok(e);
    }

    // identifier or method call
    let receiver_tok = parser.expect(TokenKind::Identifier)?;

    // method call: receiver.method(arg)
    if parser.r#match(TokenKind::Dot) {
        let method_tok = parser.expect(TokenKind::Identifier)?;
        expect_token(parser, TokenKind::LParen)?;
        let arg = parse_expr(parser)?;
        expect_token(parser, TokenKind::RParen)?;

        return Ok(Expr {
            kind: ExprKind::MethodCall(MethodCall {
                receiver: receiver_tok.lexeme,
                method: method_tok.lexeme,
                arg: Box::new(arg),
            }),
            loc: receiver_tok.loc,
        });
    }

    // variable
    Ok(Expr {
        kind: ExprKind::Variable(receiver_tok.lexeme),
        loc: receiver_tok.loc,
    })
}